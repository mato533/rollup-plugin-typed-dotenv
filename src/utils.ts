import path from 'node:path';
import fs from 'node:fs';

import dotenv from 'dotenv';

import type { DotenvParseOutput } from 'dotenv';
import type {
  EnumTypes,
  NotNullPluginOptions,
  PluginOptions,
  TypeDefinition,
} from '@/types';

export const parseOptions = (options: PluginOptions): NotNullPluginOptions => {
  const cwd = process.cwd();
  return {
    envDir: options.envDir ? path.resolve(cwd, options.envDir) : cwd,
    envPrefix: options.envPrefix ? options.envPrefix : 'ROLLUP_',
    envKey: options.envKey ? options.envKey : 'NODE_ENV',
    preventAssignment: options.preventAssignment ? true : false,
    values: options.values,
  };
};

export const getReplacements = (
  envDir: string,
  envPrefix: string,
  envKey: string,
  typeInfo: TypeDefinition,
): DotenvParseOutput => {
  const configkeys = Object.keys(typeInfo);
  const mode = process.env[envKey];
  const modeEnvFiles = mode ? [`.env.${mode}`, `.env.${mode}.local`] : [];

  return ['.env', '.env.local', ...modeEnvFiles]
    .map((envFile) => path.resolve(envDir, envFile))
    .filter((envFile) => fs.existsSync(envFile))
    .map((envFile) => fs.readFileSync(envFile, { encoding: 'utf-8' }))
    .map((envContent) => dotenv.parse(envContent))
    .map((envEntries) => {
      return Object.fromEntries(
        Object.entries(envEntries)
          .filter(([key]) => key.startsWith(envPrefix))
          .filter(([key]) => configkeys.includes(key))
          .map(([key, value]) => getReplaceKeyValue(key, value, typeInfo[key])),
      );
    })
    .reduce((pre, current) => {
      return Object.assign(pre, current);
    }, {});
};

const getReplaceKeyValue = (key: string, value: string, type: EnumTypes) => {
  return [`process.env.${key}`, perseTypedValue[type](value)];
};

const perseTypedValue = {
  string: function (value: string): string {
    return JSON.stringify(value);
  },
  number: function (value: string): string {
    return value;
  },
  boolean: function (value: string): string {
    const numbered = Number(value);
    const falsyValue = ['false', 'undefined', 'null', 'NaN'.toLowerCase(), '0'];

    return value.toLowerCase() === 'true' ||
      (numbered != 0 && !Number.isNaN(numbered)) ||
      (value.length > 0 && !falsyValue.includes(value.toLowerCase()))
      ? 'true'
      : 'false';
  },
};
