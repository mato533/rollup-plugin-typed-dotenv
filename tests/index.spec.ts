import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { rollup } from 'rollup';

import type { TypeDefinition } from '@/types';

import typedDotenv from '@/index';
import { getReplacements } from '@/utils';

interface TestParams {
  scenario: string;
  dirFixtures: string;
}

const dirnameFixtures = './__fixtures__';
const dirFixtures = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  dirnameFixtures,
);

const defineFixtureFileName = (param: TestParams) => {
  const { scenario, dirFixtures: dir } = param;
  const inputFile = path.join(dir, `${scenario}.js`);
  return inputFile;
};

const build = async (param: TestParams) => {
  const inputFile = defineFixtureFileName(param);
  const bundle = await rollup({
    input: inputFile,
    plugins: [
      typedDotenv({
        envDir: 'env',
        values: {
          ROLLUP_VALUE_STRING: 'string',
          ROLLUP_VALUE_NUMBER: 'number',
          ROLLUP_VALUE_FALSE: 'boolean',
        },
      }),
    ],
  });

  return await bundle.generate({});
};

const buildAndAssertOutput = async (param: TestParams) => {
  const output = await build(param);

  expect(output.output.length).toBe(1);
  const [{ code: generated }] = output.output;
  expect(generated).toMatchSnapshot();
};

describe('replace', () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should replace with contents of env files', async () => {
    vi.spyOn(process, 'cwd').mockReturnValue(dirFixtures);

    await buildAndAssertOutput({
      scenario: 'index',
      dirFixtures: dirFixtures,
    });
  });
});

describe('env', () => {
  const envDir = path.join(dirFixtures, 'env');
  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('should have priority order of env files', () => {
    it.each`
      key                       | expected
      ${'ROLLUP_VALUE_STRING1'} | ${'.env'}
      ${'ROLLUP_VALUE_STRING2'} | ${'.env.local'}
      ${'ROLLUP_VALUE_STRING3'} | ${'.env.test'}
      ${'ROLLUP_VALUE_STRING4'} | ${'.env.test.local'}
    `(
      'shoud be highest priority when it exists: $expected ',
      ({ key, expected }) => {
        const typeInfo: TypeDefinition = {};
        typeInfo[key] = 'string';

        const result = Object.values(
          getReplacements(envDir, 'ROLLUP_', 'NODE_ENV', typeInfo),
        );

        expect(result).toHaveLength(1);
        expect(JSON.parse(result[0])).toBe(expected);
      },
    );
  });

  it('should be enclosed in double quotes when type is string', () => {
    const typeInfo: TypeDefinition = {
      ROLLUP_VALUE_STRING: 'string',
    };
    const result = Object.values(
      getReplacements(envDir, 'ROLLUP_', 'NODE_ENV', typeInfo),
    );
    expect(result).toHaveLength(1);
    expect(result[0]).toBe(JSON.stringify('Test local value'));
  });

  it('should not be enclosed in double quotes when type is number', () => {
    const typeInfo: TypeDefinition = {
      ROLLUP_VALUE_NUMBER: 'number',
    };
    const result = Object.values(
      getReplacements(envDir, 'ROLLUP_', 'NODE_ENV', typeInfo),
    );
    expect(result).toHaveLength(1);
    expect(result[0]).toBe('2');
  });

  describe('should retern string of boolean value', () => {
    it.each`
      key            | expected
      ${'false'}     | ${false}
      ${'undefined'} | ${false}
      ${'empty'}     | ${false}
      ${'null'}      | ${false}
      ${'NaN'}       | ${false}
      ${'zero'}      | ${false}
      ${'true'}      | ${true}
      ${'one'}       | ${true}
      ${'string'}    | ${true}
    `('$key', ({ key, expected }) => {
      const typeInfo: TypeDefinition = {};
      typeInfo[`ROLLUP_VALUE_${key.toUpperCase()}`] = 'boolean';

      const result = Object.values(
        getReplacements(envDir, 'ROLLUP_', 'NODE_ENV', typeInfo),
      );

      expect(result).toHaveLength(1);
      expect(result[0]).toBe(JSON.stringify(expected));
    });
  });
});
