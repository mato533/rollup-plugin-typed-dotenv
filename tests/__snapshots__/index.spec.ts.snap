// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`option > preventAssignment 1`] = `
"/* eslint-disable no-undef */
console.log(process.env.ROLLUP_VALUE_STRING);
console.log(process.env.ROLLUP_VALUE_NUMBER);
console.log(process.env.ROLLUP_VALUE_FALSE);

process.env.ROLLUP_DEBUG = false;
if (true === true) {
  console.log(true);
}
"
`;

exports[`replace > should replace with contents of env files 1`] = `
"/* eslint-disable no-undef */
console.log(\\"Test local value\\");
console.log(2);
console.log(false);

process.env.ROLLUP_DEBUG = false;
if (process.env.ROLLUP_DEBUG === true) {
  console.log(process.env.ROLLUP_DEBUG);
}
"
`;
